// @revideo/2d/lib/utils/video/parser/sampler.js
export class FrameSampler {
    constructor(segment, sourceFps, targetFps, sum = 0) {
        this.segment = segment;
        this.sourceFps = sourceFps;
        this.targetFps = targetFps;
        this.sum = sum;
        this.framesRequested = 0;
    }
    async getNextFrame() {
        const samplingRate = this.sourceFps / this.targetFps;
        this.sum += samplingRate;
        this.framesRequested += 1;
        // Return last frame if we haven't progressed enough to get a new frame.
        if (this.sum < 1 && this.lastFrame) {
            return this.lastFrame;
        }
        // Burn frames if we have too many.
        while (this.sum >= 2) {
            // Burn frame
            const frame = await this.segment.getNextFrame();
            // If there are no more frames, return
            if (!frame) {
                return;
            }
            frame.close();
            this.sum -= 1;
        }
        // Get a new frame.
        if (this.sum >= 1 || !this.lastFrame) {
            this.sum -= 1;
            const frame = await this.segment.getNextFrame();
            // If there are no more frames, return
            if (!frame) {
                return;
            }
            this.lastFrame?.close();
            this.lastFrame = frame;
            return frame;
        }
        // One of the three if statements above is always true.
        throw new Error('Unreachable code');
    }
    getLastFrame() {
        return this.lastFrame;
    }
    getSum() {
        return this.sum;
    }
    getSegment() {
        return this.segment;
    }
    getTime(framesToSubtract) {
        return (this.framesRequested - framesToSubtract) / this.targetFps;
    }
}
