// @revideo/2d/lib/components/Media.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var Media_1;
import { DependencyContext, PlaybackState, clamp, isReactive, useLogger, useThread, } from '@revideo/core';
import { computed, initial, nodeName, signal } from '../decorators';
import { Rect } from './Rect';
const reactivePlaybackRate = `
The \`playbackRate\` of a \`Video\` cannot be reactive.

Make sure to use a concrete value and not a function:

\`\`\`ts wrong
video.playbackRate(() => 7);
\`\`\`

\`\`\`ts correct
video.playbackRate(7);
\`\`\`

If you're using a signal, extract its value before passing it to the property:

\`\`\`ts wrong
video.playbackRate(mySignal);
\`\`\`

\`\`\`ts correct
video.playbackRate(mySignal());
\`\`\`
`;
let Media = Media_1 = class Media extends Rect {
    constructor(props) {
        super(props);
        this.lastTime = -1;
        if (!this.awaitCanPlay()) {
            this.scheduleSeek(this.time());
        }
        if (props.play) {
            this.play();
        }
        this.volume = props.volume ?? 1;
        this.setVolume(this.volume);
    }
    isPlaying() {
        return this.playing();
    }
    getCurrentTime() {
        return this.clampTime(this.time());
    }
    getDuration() {
        return this.mediaElement().duration;
    }
    getVolume() {
        return this.volume;
    }
    getUrl() {
        return this.mediaElement().src;
    }
    dispose() {
        this.pause();
        this.remove();
        super.dispose();
    }
    completion() {
        return this.clampTime(this.time()) / this.getDuration();
    }
    setCurrentTime(value) {
        const media = this.mediaElement();
        if (media.readyState < 2)
            return;
        media.currentTime = value;
        this.lastTime = value;
        if (media.seeking) {
            DependencyContext.collectPromise(new Promise(resolve => {
                const listener = () => {
                    resolve();
                    media.removeEventListener('seeked', listener);
                };
                media.addEventListener('seeked', listener);
            }));
        }
    }
    setVolume(volume) {
        if (volume < 0) {
            console.warn(`volumes cannot be negative - the value will be clamped to 0.`);
        }
        const media = this.mediaElement();
        media.volume = Math.min(Math.max(volume, 0), 1);
        if (volume > 1) {
            if (this.allowVolumeAmplificationInPreview()) {
                this.amplify(media, volume);
                return;
            }
            console.warn(`you have set the volume of node ${this.key} to ${volume} - your video will be exported with the correct volume, but the browser does not support volumes higher than 1 by default. To enable volume amplification in the preview, set the "allowVolumeAmplificationInPreview" of your <Video/> or <Audio/> tag to true. Note that amplification for previews will not work if you use autoplay within the player due to browser autoplay policies: https://developer.chrome.com/blog/autoplay/#webaudio.`);
        }
    }
    amplify(node, volume) {
        const key = `${this.src()}/${this.key}`;
        if (Media_1.amplificationPool[key]) {
            Media_1.amplificationPool[key].gainNode.gain.value = volume;
            return;
        }
        const audioContext = new AudioContext();
        const sourceNode = audioContext.createMediaElementSource(node);
        const gainNode = audioContext.createGain();
        gainNode.gain.value = volume;
        sourceNode.connect(gainNode);
        gainNode.connect(audioContext.destination);
        Media_1.amplificationPool[key] = { audioContext, sourceNode, gainNode };
        if (typeof window === 'undefined' || audioContext.state !== 'suspended') {
            return;
        }
        // Start audio context after user interation, neccessary due to browser autoplay policies
        const handleInteraction = () => {
            Media_1.amplificationPool[key].audioContext.resume();
            window.removeEventListener('click', handleInteraction);
        };
        window.addEventListener('click', handleInteraction);
    }
    setPlaybackRate(playbackRate) {
        let value;
        if (isReactive(playbackRate)) {
            value = playbackRate();
            useLogger().warn({
                message: 'Invalid value set as the playback rate',
                remarks: reactivePlaybackRate,
                inspect: this.key,
                stack: new Error().stack,
            });
        }
        else {
            value = playbackRate;
        }
        this.playbackRate.context.setter(value);
        if (this.playing()) {
            if (value === 0) {
                this.pause();
            }
            else {
                const time = useThread().time;
                const start = time();
                const offset = this.time();
                this.time(() => this.clampTime(offset + (time() - start) * value));
            }
        }
    }
    scheduleSeek(time) {
        this.waitForCanPlay(this.mediaElement(), () => {
            const media = this.mediaElement();
            // Wait until the media is ready to seek again as
            // setting the time before the video doesn't work reliably.
            media.currentTime = time;
        });
    }
    /**
     * Waits for the canplay event to be fired before calling onCanPlay.
     *
     * If the media is already ready to play, onCanPlay is called immediately.
     * @param onCanPlay - The function to call when the media is ready to play.
     * @returns
     */
    waitForCanPlay(media, onCanPlay) {
        if (media.readyState >= 2) {
            onCanPlay();
            return;
        }
        const onCanPlayWrapper = () => {
            onCanPlay();
            media.removeEventListener('canplay', onCanPlayWrapper);
        };
        const onError = () => {
            const reason = this.getErrorReason(media.error?.code);
            console.log(`ERROR: Error loading video: ${this.src()}, ${reason}`);
            media.removeEventListener('error', onError);
        };
        media.addEventListener('canplay', onCanPlayWrapper);
        media.addEventListener('error', onError);
    }
    /**
     * Returns true if we should wait for the media to be ready to play.
     */
    waitForCanPlayNecessary(media) {
        if (media.readyState >= 2) {
            return false;
        }
        return (this.awaitCanPlay() ||
            this.view().playbackState() === PlaybackState.Rendering);
    }
    play() {
        const time = useThread().time;
        const start = time();
        const offset = this.time();
        const playbackRate = this.playbackRate();
        this.playing(true);
        this.time(() => this.clampTime(offset + (time() - start) * playbackRate));
    }
    pause() {
        this.playing(false);
        this.time.save();
        this.mediaElement().pause();
    }
    clampTime(time) {
        const duration = this.getDuration();
        if (this.loop()) {
            time %= duration;
        }
        return clamp(0, duration, time);
    }
    collectAsyncResources() {
        super.collectAsyncResources();
        this.seekedMedia();
    }
    getErrorReason(errCode) {
        let reason;
        switch (errCode) {
            case 1:
                reason = 'MEDIA_ERR_ABORTED';
                break;
            case 2:
                reason = 'MEDIA_ERR_NETWORK. This might be a 404 error.';
                break;
            case 3:
                reason =
                    'MEDIA_ERR_DECODE. This might be an issue with your video file.';
                break;
            case 4:
                reason =
                    'MEDIA_ERR_SRC_NOT_SUPPORTED. If you are sure that the path to the video is correct, this might be a CORS error.';
                break;
            default:
                reason = 'UNKNOWN';
        }
        return reason;
    }
};
Media.amplificationPool = {};
__decorate([
    signal()
], Media.prototype, "src", void 0);
__decorate([
    initial(false),
    signal()
], Media.prototype, "loop", void 0);
__decorate([
    initial(1),
    signal()
], Media.prototype, "playbackRate", void 0);
__decorate([
    initial(0),
    signal()
], Media.prototype, "time", void 0);
__decorate([
    initial(false),
    signal()
], Media.prototype, "playing", void 0);
__decorate([
    initial(true),
    signal()
], Media.prototype, "awaitCanPlay", void 0);
__decorate([
    initial(false),
    signal()
], Media.prototype, "allowVolumeAmplificationInPreview", void 0);
__decorate([
    computed()
], Media.prototype, "completion", null);
__decorate([
    computed()
], Media.prototype, "amplify", null);
Media = Media_1 = __decorate([
    nodeName('Media')
], Media);
export { Media };
