// @revideo/2d/lib/components/TxtLeaf.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var TxtLeaf_1;
import { BBox, capitalize, lazy, textLerp } from '@revideo/core';
import { computed, initial, interpolation, nodeName, signal, } from '../decorators';
import { Shape } from './Shape';
export const TXT_TYPE = Symbol('Txt');
let TxtLeaf = TxtLeaf_1 = class TxtLeaf extends Shape {
    constructor({ children, ...rest }) {
        super(rest);
        if (children) {
            this.text(children);
        }
    }
    parentTxt() {
        const parent = this.parent();
        if (!parent) {
            return null;
        }
        if (!(TXT_TYPE in parent)) {
            return null;
        }
        return parent;
    }
    async draw(context) {
        this.requestFontUpdate();
        this.applyStyle(context);
        await document.fonts?.ready;
        this.applyText(context);
        context.font = this.styles.font;
        context.textBaseline = 'bottom';
        if ('letterSpacing' in context) {
            context.letterSpacing = `${this.letterSpacing()}px`;
        }
        const fontOffset = context.measureText('').fontBoundingBoxAscent;
        const parentRect = this.element.getBoundingClientRect();
        const { width, height } = this.size();
        const range = document.createRange();
        let line = '';
        const lineRect = new BBox();
        for (const childNode of this.element.childNodes) {
            if (!childNode.textContent) {
                continue;
            }
            range.selectNodeContents(childNode);
            const rangeRect = range.getBoundingClientRect();
            const x = width / -2 + rangeRect.left - parentRect.left;
            const y = height / -2 + rangeRect.top - parentRect.top + fontOffset;
            if (lineRect.y === y) {
                lineRect.width += rangeRect.width;
                line += childNode.textContent;
            }
            else {
                await this.drawText(context, line, lineRect);
                lineRect.x = x;
                lineRect.y = y;
                lineRect.width = rangeRect.width;
                lineRect.height = rangeRect.height;
                line = childNode.textContent;
            }
        }
        await this.drawText(context, line, lineRect);
    }
    async drawText(context, text, box) {
        const y = box.y;
        text = text.replace(/\s+/g, ' ');
        if (this.lineWidth() <= 0) {
            context.fillText(text, box.x, y);
        }
        else if (this.strokeFirst()) {
            context.strokeText(text, box.x, y);
            context.fillText(text, box.x, y);
        }
        else {
            context.fillText(text, box.x, y);
            context.strokeText(text, box.x, y);
        }
    }
    getCacheBBox() {
        const size = this.computedSize();
        const range = document.createRange();
        range.selectNodeContents(this.element);
        const bbox = range.getBoundingClientRect();
        const lineWidth = this.lineWidth();
        // We take the default value of the miterLimit as 10.
        const miterLimitCoefficient = this.lineJoin() === 'miter' ? 0.5 * 10 : 0.5;
        return new BBox(-size.width / 2, -size.height / 2, bbox.width, bbox.height)
            .expand([0, this.fontSize() * 0.5])
            .expand(lineWidth * miterLimitCoefficient);
    }
    applyFlex() {
        super.applyFlex();
        this.element.style.display = 'inline';
    }
    updateLayout() {
        this.applyFont();
        this.applyFlex();
        // Make sure the text is aligned correctly even if the text is smaller than
        // the container.
        if (this.justifyContent.isInitial()) {
            this.element.style.justifyContent =
                this.styles.getPropertyValue('text-align');
        }
        const wrap = this.styles.whiteSpace !== 'nowrap' && this.styles.whiteSpace !== 'pre';
        if (wrap) {
            this.element.innerText = '';
            if (TxtLeaf_1.segmenter) {
                for (const word of TxtLeaf_1.segmenter.segment(this.text())) {
                    this.element.appendChild(document.createTextNode(word.segment));
                }
            }
            else {
                for (const word of this.text().split('')) {
                    this.element.appendChild(document.createTextNode(word));
                }
            }
        }
        else if (this.styles.whiteSpace === 'pre') {
            this.element.innerText = '';
            for (const line of this.text().split('\n')) {
                this.element.appendChild(document.createTextNode(line + '\n'));
            }
        }
        else {
            this.element.innerText = this.text();
        }
    }
};
__decorate([
    initial(''),
    interpolation(textLerp),
    signal()
], TxtLeaf.prototype, "text", void 0);
__decorate([
    computed()
], TxtLeaf.prototype, "parentTxt", null);
__decorate([
    lazy(() => {
        try {
            return new Intl.Segmenter(undefined, {
                granularity: 'grapheme',
            });
        }
        catch (e) {
            return null;
        }
    })
], TxtLeaf, "segmenter", void 0);
TxtLeaf = TxtLeaf_1 = __decorate([
    nodeName('TxtLeaf')
], TxtLeaf);
export { TxtLeaf };
/**
 * Overwrite all getters for signal values to return the parent value if it
 * exists.
 *
 * The getters on the TxtLeaf class are used by the `@signal` decorators and
 * are not used by the class or its consumers directly.
 *
 * Check out 2d/src/lib/utils/makeSignalExtensions.ts if this is confusing.
 */
[
    'fill',
    'stroke',
    'lineWidth',
    'strokeFirst',
    'lineCap',
    'lineJoin',
    'lineDash',
    'lineDashOffset',
].forEach(prop => {
    TxtLeaf.prototype[`get${capitalize(prop)}`] = function () {
        return (this.parentTxt()?.[prop]() ??
            this[prop].context.getInitial());
    };
});
