// @revideo/vite-plugin/lib/partials/projects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.projectsPlugin = projectsPlugin;
function projectsPlugin({ buildForEditor, projects, }) {
    // TODO(refactor): use version information
    // const versions = JSON.stringify(getVersions());
    return {
        name: 'revideo:project',
        config(config) {
            return {
                build: {
                    target: buildForEditor ? 'esnext' : 'modules',
                    assetsDir: './',
                    rollupOptions: {
                        preserveEntrySignatures: 'strict',
                        input: Object.fromEntries(projects.list.map(project => [project.name, project.url])),
                    },
                },
                server: {
                    port: config?.server?.port ?? 9000,
                },
                esbuild: {
                    jsx: 'automatic',
                    jsxImportSource: '@revideo/2d/lib',
                },
                optimizeDeps: {
                    entries: projects.list.map(project => project.url),
                    exclude: ['preact', 'preact/*', '@preact/signals'],
                },
            };
        },
    };
}
