// @revideo/core/lib/types/Color.js
import { Color as ChromaColor, mix } from 'chroma-js';
import { SignalContext } from '../signals';
// iife prevents tree shaking from stripping our methods.
const ExtendedColor = (() => {
    ChromaColor.symbol = ChromaColor.prototype.symbol = Symbol.for('@revideo/core/types/Color');
    ChromaColor.lerp = ChromaColor.prototype.lerp = (from, to, value, colorSpace = 'lch') => {
        if (typeof from === 'string') {
            from = new ChromaColor(from);
        }
        if (typeof to === 'string') {
            to = new ChromaColor(to);
        }
        const fromIsColor = from instanceof ChromaColor;
        const toIsColor = to instanceof ChromaColor;
        if (!fromIsColor) {
            from = toIsColor
                ? to.alpha(0)
                : new ChromaColor('rgba(0, 0, 0, 0)');
        }
        if (!toIsColor) {
            to = fromIsColor
                ? from.alpha(0)
                : new ChromaColor('rgba(0, 0, 0, 0)');
        }
        return mix(from, to, value, colorSpace);
    };
    ChromaColor.createLerp = ChromaColor.prototype.createLerp =
        (colorSpace) => (from, to, value) => ChromaColor.lerp(from, to, value, colorSpace);
    ChromaColor.createSignal = (initial, interpolation = ChromaColor.lerp) => {
        return new SignalContext(initial, interpolation, undefined, value => new ChromaColor(value)).toSignal();
    };
    ChromaColor.prototype.toSymbol = () => {
        return ChromaColor.symbol;
    };
    ChromaColor.prototype.toUniform = function (gl, location) {
        gl.uniform4fv(location, this.gl());
    };
    ChromaColor.prototype.serialize = function () {
        return this.css();
    };
    ChromaColor.prototype.lerp = function (to, value, colorSpace) {
        return ChromaColor.lerp(this, to, value, colorSpace);
    };
    return ChromaColor;
})();
export { ExtendedColor as Color };
