// @revideo/ffmpeg/dist/ffmpeg-exporter-server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FFmpegExporterServer = exports.extensions = void 0;
const telemetry_1 = require("@revideo/telemetry");
const ffmpeg = require("fluent-ffmpeg");
const os = require("os");
const path = require("path");
const image_stream_1 = require("./image-stream");
const settings_1 = require("./settings");
const pixelFormats = {
    mp4: 'yuv420p',
    webm: 'yuva420p',
    proRes: 'yuva444p10le',
};
exports.extensions = {
    mp4: 'mp4',
    webm: 'webm',
    proRes: 'mov',
};
/**
 * The server-side implementation of the FFmpeg video exporter.
 */
class FFmpegExporterServer {
    constructor(settings) {
        if (settings.exporter.name !== '@revideo/core/ffmpeg') {
            throw new Error('Invalid exporter');
        }
        this.settings = settings;
        this.format = settings.exporter.options.format;
        this.jobFolder = path.join(os.tmpdir(), `revideo-${this.settings.name}-${settings.hiddenFolderId}`);
        this.stream = new image_stream_1.ImageStream();
        ffmpeg.setFfmpegPath(settings_1.ffmpegSettings.getFfmpegPath());
        this.command = ffmpeg();
        // Input image sequence
        this.command
            .input(this.stream)
            .inputFormat('image2pipe')
            .inputFps(settings.fps);
        // Output settings
        const size = {
            x: Math.round(settings.size.x * settings.resolutionScale),
            y: Math.round(settings.size.y * settings.resolutionScale),
        };
        this.command
            .output(path.join(this.jobFolder, `visuals.${exports.extensions[this.format]}`))
            .outputOptions([`-pix_fmt ${pixelFormats[this.format]}`, '-shortest'])
            .outputFps(settings.fps)
            .size(`${size.x}x${size.y}`);
        if (this.format === 'proRes') {
            this.command.outputOptions(['-c:v prores_ks', '-profile:v 4444']);
        }
        this.command.outputOptions(['-movflags +faststart']);
        this.promise = new Promise((resolve, reject) => {
            this.command.on('end', resolve).on('error', reject);
        });
    }
    async start() {
        this.command.run();
    }
    async handleFrame({ data }) {
        const base64Data = data.slice(data.indexOf(',') + 1);
        this.stream.pushImage(Buffer.from(base64Data, 'base64'));
    }
    async end(result) {
        this.stream.pushImage(null);
        if (result === 1) {
            try {
                this.command.kill('SIGKILL');
                await this.promise;
            }
            catch (err) {
                (0, telemetry_1.sendEvent)(telemetry_1.EventName.Error, { message: err.message });
            }
        }
        else {
            await this.promise;
        }
    }
    async kill() {
        try {
            this.command.kill('SIGKILL');
            await this.promise;
        }
        catch (_) {
            return;
        }
    }
}
exports.FFmpegExporterServer = FFmpegExporterServer;
