// @revideo/ffmpeg/dist/video-frame-extractor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoFrameExtractor = void 0;
const telemetry_1 = require("@revideo/telemetry");
const ffmpeg = require("fluent-ffmpeg");
const fs = require("fs");
const os = require("os");
const path = require("path");
const uuid_1 = require("uuid");
const settings_1 = require("./settings");
const utils_1 = require("./utils");
/**
 * Walks through a video file and extracts frames.
 */
class VideoFrameExtractor {
    constructor(filePath, startTime, fps, duration) {
        this.ffmpegPath = settings_1.ffmpegSettings.getFfmpegPath();
        this.buffer = Buffer.alloc(0);
        this.bufferOffset = 0;
        // Images are buffered in memory until they are requested.
        this.imageBuffers = [];
        this.lastImage = null;
        this.framesProcessed = 0;
        this.width = 0;
        this.height = 0;
        this.frameSize = 0;
        this.codec = null;
        this.process = null;
        this.terminated = false;
        this.state = 'processing';
        this.filePath = filePath;
        this.downloadedFilePath = VideoFrameExtractor.downloadedVideoMap.get(filePath)?.localPath;
        this.startTimeOffset = VideoFrameExtractor.downloadedVideoMap.get(filePath)
            ?.startTimeOffset;
        this.startTime = startTime;
        this.duration = duration;
        this.toTime = this.getEndTime(this.startTime);
        this.fps = fps;
        (0, utils_1.getVideoMetadata)(this.downloadedFilePath).then(metadata => {
            this.width = metadata.width;
            this.height = metadata.height;
            this.frameSize = this.width * this.height * 4;
            this.buffer = Buffer.alloc(this.frameSize);
            this.codec = metadata.codec;
            if (this.startTime >= this.duration) {
                this.process = this.createFfmpegProcessToExtractFirstFrame(this.downloadedFilePath, this.codec);
                return;
            }
            this.process = this.createFfmpegProcess(this.startTime - this.startTimeOffset, this.toTime, this.downloadedFilePath, this.fps, this.codec);
        });
    }
    static downloadVideoChunk(url, startTime, endTime) {
        const outputDir = path.join(os.tmpdir(), `revideo-decoder-chunks`);
        if (!fs.existsSync(outputDir)) {
            fs.mkdirSync(outputDir, { recursive: true });
        }
        return new Promise((resolve, reject) => {
            ffmpeg.ffprobe(url, (err, metadata) => {
                if (err) {
                    reject(err);
                    return;
                }
                const format = metadata.format.format_name?.split(',')[-1] || 'mp4';
                const outputFileName = `chunk_${(0, uuid_1.v4)()}.${format}`;
                const outputPath = path.join(outputDir, outputFileName);
                const toleranceInSeconds = 0.5;
                const adjustedStartTime = Math.max(startTime - toleranceInSeconds, 0);
                ffmpeg(url)
                    .setFfmpegPath(settings_1.ffmpegSettings.getFfmpegPath())
                    .inputOptions([
                    `-ss ${adjustedStartTime}`,
                    `-to ${endTime + toleranceInSeconds}`,
                ])
                    .outputOptions(['-c copy'])
                    .output(outputPath)
                    .on('end', () => {
                    this.downloadedVideoMap.set(url, {
                        localPath: outputPath,
                        startTimeOffset: adjustedStartTime,
                    });
                    resolve(outputPath);
                })
                    .on('error', err => reject(err))
                    .run();
            });
        });
    }
    getTime() {
        return this.startTime + this.framesProcessed / this.fps;
    }
    getLastTime() {
        return this.startTime + (this.framesProcessed - 1) / this.fps;
    }
    getLastFrame() {
        return this.lastImage;
    }
    getWidth() {
        return this.width;
    }
    getHeight() {
        return this.height;
    }
    getEndTime(startTime) {
        return Math.min(startTime + VideoFrameExtractor.chunkLengthInSeconds, this.duration);
    }
    getArgs(codec, range, fps) {
        const inputOptions = [];
        const outputOptions = [];
        inputOptions.push('-loglevel', settings_1.ffmpegSettings.getLogLevel());
        if (range) {
            inputOptions.push(...['-ss', range[0].toFixed(2), '-to', range[1].toFixed(2)]);
        }
        if (codec === 'vp9') {
            inputOptions.push('-vcodec', 'libvpx-vp9');
        }
        if (fps) {
            outputOptions.push('-vf', `fps=fps=${fps}`);
        }
        if (!range) {
            outputOptions.push('-vframes', '1');
        }
        outputOptions.push('-f', 'rawvideo');
        outputOptions.push('-pix_fmt', 'rgba');
        return { inputOptions, outputOptions };
    }
    createFfmpegProcess(startTime, toTime, filePath, fps, codec) {
        const { inputOptions, outputOptions } = this.getArgs(codec, [startTime, toTime], fps);
        const process = ffmpeg(filePath)
            .setFfmpegPath(this.ffmpegPath)
            .inputOptions(inputOptions)
            .outputOptions(outputOptions)
            .on('end', () => {
            this.handleClose(0);
        })
            .on('error', err => {
            this.handleError(err);
        })
            .on('stderr', stderrLine => {
            console.log(stderrLine);
        })
            .on('stdout', stderrLine => {
            console.log(stderrLine);
        });
        const ffstream = process.pipe();
        ffstream.on('data', (data) => {
            this.processData(data);
        });
        return process;
    }
    /**
     * We call this in the case that the time requested is greater than the
     * duration of the video. In this case, we want to display the first frame
     * of the video.
     *
     * Note: This does NOT match the behavior of the old implementation
     * inside of 2d/src/lib/components/Video.ts. In the old implementation, the
     * last frame is shown instead of the first frame.
     */
    createFfmpegProcessToExtractFirstFrame(filePath, codec) {
        const { inputOptions, outputOptions } = this.getArgs(codec, undefined, undefined);
        const process = ffmpeg(filePath)
            .setFfmpegPath(this.ffmpegPath)
            .inputOptions(inputOptions)
            .outputOptions(outputOptions)
            .on('end', () => {
            this.handleClose(0);
        })
            .on('error', err => {
            this.handleError(err);
        })
            .on('stderr', stderrLine => {
            console.log(stderrLine);
        })
            .on('stdout', stderrLine => {
            console.log(stderrLine);
        });
        const ffstream = process.pipe();
        ffstream.on('data', (data) => {
            this.processData(data);
        });
        return process;
    }
    processData(data) {
        let dataOffset = 0;
        while (dataOffset < data.length) {
            const remainingSpace = this.frameSize - this.bufferOffset;
            const chunkSize = Math.min(remainingSpace, data.length - dataOffset);
            data.copy(this.buffer, this.bufferOffset, dataOffset, dataOffset + chunkSize);
            this.bufferOffset += chunkSize;
            dataOffset += chunkSize;
            // We have a complete frame
            if (this.bufferOffset === this.frameSize) {
                this.imageBuffers.push(Buffer.from(this.buffer)); // Create a copy
                this.bufferOffset = 0; // Reset buffer for next frame
            }
        }
    }
    async popImage() {
        if (this.imageBuffers.length) {
            const image = this.imageBuffers.shift();
            this.framesProcessed++;
            this.lastImage = image;
            return image;
        }
        if (this.state === 'error') {
            throw new Error('An error occurred while extracting the video frames.');
        }
        // If the video is done and there are no more frames to extract, return the last frame.
        if (this.state === 'done' && this.toTime >= this.duration) {
            return this.lastImage;
        }
        // If there are more frames to extract, request the next chunk.
        if (this.state === 'done') {
            this.startTime = this.toTime;
            this.toTime = Math.min(this.startTime + VideoFrameExtractor.chunkLengthInSeconds, this.duration);
            if (!this.codec) {
                throw new Error("Can't extract frames without a codec. This error should never happen.");
            }
            this.process = this.createFfmpegProcess(this.startTime, this.toTime, this.downloadedFilePath, this.fps, this.codec);
            this.state = 'processing';
        }
        while (this.imageBuffers.length < 1) {
            await new Promise(resolve => setTimeout(resolve, 50));
        }
        const image = this.imageBuffers.shift();
        this.framesProcessed++;
        this.lastImage = image;
        return image;
    }
    handleClose(code) {
        this.state = code === 0 ? 'done' : 'error';
    }
    async handleError(err) {
        const code = err.code;
        if (this.terminated) {
            return;
        }
        if (code === 'ENOENT') {
            (0, telemetry_1.sendEvent)(telemetry_1.EventName.Error, { error: 'ffmpeg-not-found' });
            throw new Error('Error: ffmpeg not found. Make sure ffmpeg is installed on your system.');
        }
        else if (err.message.includes('SIGSEGV')) {
            (0, telemetry_1.sendEvent)(telemetry_1.EventName.Error, {
                error: 'ffmpeg-sigsegv',
                message: err.message,
            });
            throw new Error(`Error: Segmentation fault when running ffmpeg. This is a common issue on Linux, you might be able to fix it by installing nscd ('sudo apt-get install nscd'). For more information, see https://docs.re.video/common-issues/ffmpeg/`);
        }
        else {
            await (0, telemetry_1.sendEvent)(telemetry_1.EventName.Error, {
                error: 'ffmpeg-error',
                message: err.message,
            });
            throw new Error(`An ffmpeg error occurred while fetching frames from source video ${this.filePath}: ${err}`);
        }
    }
    destroy() {
        this.terminated = true;
        this.process?.kill('SIGTERM');
    }
}
exports.VideoFrameExtractor = VideoFrameExtractor;
VideoFrameExtractor.chunkLengthInSeconds = 5;
VideoFrameExtractor.downloadedVideoMap = new Map();
