// @revideo/core/lib/threading/Thread.js
import { decorate, threadable } from '../decorators';
import { getTaskName, setTaskName } from '../flow';
import { createSignal } from '../signals';
import { endThread, startThread, useLogger } from '../utils';
import { isThreadGenerator } from './ThreadGenerator';
const reusedGenerator = `
This usually happens when you mistakenly reuse a generator that is already
running.

For example, using \`yield\` here will run the opacity generator concurrently and
store it in the \`task\` variable (in case you want to cancel or await it later):

\`\`\`ts
const task = yield rect().opacity(1, 1);
\`\`\`

Trying to \`yield\` this task again will cause the current error:

\`\`\`ts
yield task;
\`\`\`

Passing it to other flow functions will also cause the error:

\`\`\`ts
// prettier-ignore
yield* all(task);
\`\`\`

Try to investigate your code looking for \`yield\` statements whose return value
is reused in this way. Here's an example of a common mistake:

\`\`\`ts wrong
// prettier-ignore
yield* all(
  yield rect().opacity(1, 1), 
  yield rect().x(200, 1),
);
\`\`\`

\`\`\`ts correct
// prettier-ignore
yield* all(
  rect().opacity(1, 1), 
  rect().x(200, 1),
);
\`\`\`
`;
decorate(noop, threadable());
export function* noop() {
    // do nothing
}
/**
 * A class representing an individual thread.
 *
 * @remarks
 * Thread is a wrapper for a generator that can be executed concurrently.
 *
 * Aside from the main thread, all threads need to have a parent.
 * If a parent finishes execution, all of its child threads are terminated.
 */
export class Thread {
    /**
     * The fixed time of this thread.
     *
     * @remarks
     * Fixed time is a multiple of the frame duration. It can be used to account
     * for the difference between this thread's {@link time} and the time of the
     * current animation frame.
     */
    get fixed() {
        return this.fixedTime;
    }
    /**
     * Check if this thread or any of its ancestors has been canceled.
     */
    get canceled() {
        return this.isCanceled || (this.parent?.canceled ?? false);
    }
    get paused() {
        return this.isPaused || (this.parent?.paused ?? false);
    }
    constructor(
    /**
     * The generator wrapped by this thread.
     */
    runner) {
        this.runner = runner;
        this.children = [];
        /**
         * The current time of this thread.
         *
         * @remarks
         * Used by {@link flow.waitFor} and other time-based functions to properly
         * support durations shorter than one frame.
         */
        this.time = createSignal(0);
        this.parent = null;
        this.isCanceled = false;
        this.isPaused = false;
        this.fixedTime = 0;
        this.queue = [];
        if (this.runner.task) {
            useLogger().error({
                message: `The generator "${getTaskName(this.runner)}" is already being executed by another thread.`,
                remarks: reusedGenerator,
            });
            this.runner = noop();
        }
        this.runner.task = this;
    }
    /**
     * Progress the wrapped generator once.
     */
    next() {
        if (this.paused) {
            return {
                value: null,
                done: false,
            };
        }
        startThread(this);
        const result = this.runner.next(this.value);
        endThread(this);
        this.value = null;
        return result;
    }
    /**
     * Prepare the thread for the next update cycle.
     *
     * @param dt - The delta time of the next cycle.
     */
    update(dt) {
        if (!this.paused) {
            this.time(this.time() + dt);
            this.fixedTime += dt;
        }
        this.children = this.children.filter(child => !child.canceled);
    }
    spawn(child) {
        if (!isThreadGenerator(child)) {
            child = child();
        }
        this.queue.push(child);
        return child;
    }
    add(child) {
        child.parent = this;
        child.isCanceled = false;
        child.time(this.time());
        child.fixedTime = this.fixedTime;
        this.children.push(child);
        setTaskName(child.runner, `unknown ${this.children.length}`);
    }
    drain(callback) {
        this.queue.forEach(callback);
        this.queue = [];
    }
    cancel() {
        this.runner.return();
        this.isCanceled = true;
        this.parent = null;
        this.drain(task => task.return());
    }
    pause(value) {
        this.isPaused = value;
    }
}
