// @revideo/core/lib/exporter/WasmExporter.js
import loadMp4Module from 'mp4-wasm';
import { download } from './download-videos';
export class WasmExporter {
    static async create(project, settings) {
        return new WasmExporter(project, settings);
    }
    constructor(project, settings) {
        this.project = project;
        this.settings = settings;
    }
    async start() {
        const resp = await fetch('/@mp4-wasm');
        const buffer = await resp.arrayBuffer();
        const mp4 = await loadMp4Module({
            wasmBinary: buffer,
        });
        this.encoder = mp4.createWebCodecsEncoder({
            width: this.settings.size.x,
            height: this.settings.size.y,
            fps: this.project.settings.rendering.fps,
        });
    }
    async handleFrame(canvas) {
        const frame = new VideoFrame(canvas, { timestamp: 0 });
        await this.encoder.addFrame(frame);
        frame.close();
    }
    async stop() {
        const buf = await this.encoder.end();
        const formData = new FormData();
        formData.append('file', new Blob([buf], { type: 'video/mp4' }), 'video.mp4');
        formData.append('tempDir', `revideo-${this.settings.name}-${this.settings.hiddenFolderId}`);
        await fetch('/revideo-ffmpeg-decoder/finished', {
            method: 'POST',
            headers: {
                // eslint-disable-next-line
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({}),
        });
        await fetch('/uploadVideoFile', {
            method: 'POST',
            body: formData,
        });
    }
    async generateAudio(assets, startFrame, endFrame) {
        await fetch('/audio-processing/generate-audio', {
            method: 'POST',
            // TODO: add type and validate on the other side
            body: JSON.stringify({
                tempDir: `revideo-${this.settings.name}-${this.settings.hiddenFolderId}`,
                assets,
                startFrame,
                endFrame,
                fps: this.settings.fps,
            }),
        });
    }
    async mergeMedia() {
        const outputFilename = this.settings.name;
        const tempDir = `revideo-${this.settings.name}-${this.settings.hiddenFolderId}`;
        await fetch('/audio-processing/merge-media', {
            method: 'POST',
            // TODO: add type and validate on the other side
            body: JSON.stringify({
                outputFilename,
                tempDir,
                format: 'mp4',
            }),
        });
    }
    async downloadVideos(assets) {
        await download(assets);
    }
}
WasmExporter.id = '@revideo/core/wasm';
WasmExporter.displayName = 'Video (Wasm)';
